### PR TITLE
Improvements to the http operation parameter validation(Duplicate body, duplicate types on params)

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix https://github.com/Azure/cadl-azure/issues/1057

Changes:
-  Add some new logic in rest library so it deals with the logic of understanding operation parameters and emitting errors. There is no point waiting until the emitter to figure out this is an invalid rest api spec.
- Validate for uses of multiple param type decorator `@header` `@path`, etc.

The changes to the generated OpenAPI seems to be because we were completely ignoring the params without any annotation before(Might have been a recent regression)

### CODE REVIEW

File: packages/samples/test/output/nested/openapi.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
   "tags": [],
   "paths": {
     "/sub/a/subsub": {
-      "get": {
+      "post": {
```
Comment (timotheeguerin) [inline packages/samples/test/output/nested/openapi.json:10]: The change of behavior here is that it was before using the following default
| Case                 | Before  | Now     |
| -------------------- | ------- | ------- |
| body with @body      | `@post` | `@post` |
| body with unannotted | `@get`  | `@post` |
| no body              | `@get`  | `@get`  |

File: packages/rest/test/test-routes.ts [hunk 2]
```diff
@@ -169,4 +169,88 @@ describe("rest: routes", () => {
       },
     ]);
   });
+
+  describe("operation parameters", () => {
+    it("emit diagnostic when there is multiple unannotted parameters", async () => {
```
Comment (markcowl) [inline packages/rest/test/test-routes.ts:174]:
```suggestion
    it("emit diagnostic for parameters with multiple http request annotations", async () => {
```

File: packages/rest/test/test-routes.ts [hunk 3]
```diff
@@ -169,4 +169,88 @@ describe("rest: routes", () => {
...
+        @route("/test")
+        @get op get(@body body: string, @path @query multiParam: string): string;
+      `);
+
+      strictEqual(diagnostics.length, 1);
+      strictEqual(diagnostics[0].code, "@cadl-lang/rest/operation-param-duplicate-type");
+      strictEqual(diagnostics[0].message, "Param multiParam has multiple types: [query, path]");
+    });
+
+    it("emit diagnostic when there is multiple unannotted parameters", async () => {
```
Comment (markcowl) [inline packages/rest/test/test-routes.ts:185]:
```suggestion
    it("emit diagnostic when there are multiple unannotated parameters", async () => {
```

File: packages/rest/test/test-routes.ts [hunk 4]
```diff
@@ -169,4 +169,88 @@ describe("rest: routes", () => {
...
+
+      strictEqual(diagnostics.length, 1);
+      strictEqual(diagnostics[0].code, "@cadl-lang/rest/duplicate-body");
+      strictEqual(
+        diagnostics[0].message,
+        "Operation has multiple unannotated parameters. There can only be one representing the body"
+      );
+    });
+
+    it("emit diagnostic when there is an unannotted parameter but a @body param", async () => {
```
Comment (markcowl) [inline packages/rest/test/test-routes.ts:199]:
```suggestion
    it("emit diagnostic when there is an unannotated parameter and a @body param", async () => {
```

File: packages/rest/test/test-routes.ts [hunk 5]
```diff
@@ -169,4 +169,88 @@ describe("rest: routes", () => {
...
+
+      strictEqual(diagnostics.length, 1);
+      strictEqual(diagnostics[0].code, "@cadl-lang/rest/duplicate-body");
+      strictEqual(
+        diagnostics[0].message,
+        "Operation has a @body and an unannotated parameter. There can only be one representing the body"
+      );
+    });
+
+    it("emit diagnostic when there is multiple @body param", async () => {
```
Comment (markcowl) [inline packages/rest/test/test-routes.ts:213]:
```suggestion
    it("emit diagnostic when there are multiple @body param", async () => {
```

File: packages/rest/test/test-routes.ts [hunk 6]
```diff
@@ -169,4 +169,88 @@ describe("rest: routes", () => {
...
+      deepStrictEqual(routes, [
+        {
+          verb: "get",
+          path: "/test",
+          params: { params: [{ type: "query", name: "select" }], body: "bodyParam" },
+        },
+      ]);
+    });
+
+    it("resolve body when only unannoted parameter", async () => {
```
Comment (markcowl) [inline packages/rest/test/test-routes.ts:240]:
```suggestion
    it("resolves a single unannotated parameter as request body", async () => {
```

File: packages/samples/test/output/grpc-library-example/openapi.json [hunk 7]
```diff
@@ -185,7 +185,17 @@
         },
         "tags": [
           "LibraryService"
-        ]
+        ],
+        "requestBody": {
```
Comment (markcowl) [inline packages/samples/test/output/grpc-library-example/openapi.json:189]: wow, how did we miss this?

File: packages/samples/test/output/grpc-kiosk-example/openapi.json [hunk 8]
```diff
@@ -318,7 +318,21 @@
         },
         "tags": [
           "Display"
-        ]
+        ],
+        "requestBody": {
```
Comment (markcowl) [inline packages/samples/test/output/grpc-kiosk-example/openapi.json:322]: Hmmm, not sure how we missed this before

File: packages/samples/test/output/nested/openapi.json (hunk 1 above)
Comment (markcowl) [reply packages/samples/test/output/nested/openapi.json:10]: I think, if we are going to default operations to `@get`, then we should throw if there is a body or an unannotated parameter.  Per discussion, I believe we should not be doing operation guessing at all, and require specs using the openapi generators to specify verbs for all operations.  Similarly, an operation marked with `@get` should throw if there are any unannotated or body parameters, probably at the http level.

**Review by markcowl — CHANGES_REQUESTED** [review]
This looks good, a couple of minor nits and then I think we need to get away from guessing http operation verbs.

File: packages/samples/test/output/grpc-library-example/openapi.json (hunk 7 above)
Comment (timotheeguerin) [reply packages/samples/test/output/grpc-library-example/openapi.json:189]: probably happened during the migration to interface and there was too much noise to notice

File: packages/samples/test/output/nested/openapi.json (hunk 1 above)
Comment (timotheeguerin) [reply packages/samples/test/output/nested/openapi.json:10]: on a related note, I see we still default the verb from the name of the operation, thought we agreed to remove that for now. Should I update this as well then?

Comment (markcowl) [reply packages/samples/test/output/nested/openapi.json:10]: I think that's right, we should not have http verb inference in our openapi and code emitters, but I would want to be sure @bterlson is on the same page about this.  At the very least, if we are going to do any verb inference, it should eb simple, and should not change based on presence or absence of body (or unannotated) parameters.

Comment (bterlson) [reply packages/samples/test/output/nested/openapi.json:10]: Yep, fine dropping inference based on operation name for now.

**Review by daviwil — APPROVED** [review]
Great fixes, thanks!

**Review by markcowl — APPROVED** [review]

**Pull request merged**